File: impls/reference/wash/executor.py
"""Command execution: interpreter resolution and pipeline evaluation."""

from __future__ import annotations

import fnmatch
import os
import selectors
import subprocess
from dataclasses import dataclass
from pathlib import Path

from wash.filesystem import implied_cat_bytes
from wash.metadata import CommandMetadata, map_exit_status
from wash.parser import CommandStage, PipelineParse, RawCommandParse


class ExecutionError(Exception):
    def __init__(self, message: str, *, status: int = 500) -> None:
        super().__init__(message)
        self.message = message
        self.status = status


@dataclass
class ExecRule:
    pattern: str
    interpreter: str
    extra_args: list[str]


@dataclass
class StageResult:
    name: str
    exit_code: int
    stdout: bytes
    stderr: bytes
    http_status: int


@dataclass
class PipelineResult:
    stdout: bytes
    stderr: bytes
    content_type: str
    stages: list[StageResult]
    pipeline_description: str
    source_path: str | None
    final_command: str | None = None
    failing_stage: StageResult | None = None
    http_status: int = 200


@dataclass
class ExecConfig:
    rules: list[ExecRule]
    malformed: bool = False
    malformed_reason: str | None = None


def load_exec_rules(root: Path) -> ExecConfig:
    exec_file = root / "exec"
    rules: list[ExecRule] = []
    if not exec_file.is_file():
        return ExecConfig(rules)
    for line in exec_file.read_text(encoding="utf-8").splitlines():
        stripped = line.strip()
        if not stripped or stripped[0] == "#":
            continue
        parts = stripped.split()
        if len(parts) < 2:
            return ExecConfig(
                rules,
                malformed=True,
                malformed_reason=f"malformed exec rule: {line!r}",
            )
        rules.append(
            ExecRule(
                pattern=parts[0],
                interpreter=parts[1],
                extra_args=parts[2:],
            )
        )
    return ExecConfig(rules)


def _has_glob(pattern: str) -> bool:
    return any(ch in pattern for ch in "*?[]")


def _rule_matches(rule: ExecRule, command_path: Path, command_dir: Path) -> bool:
    basename = command_path.name
    if not _has_glob(rule.pattern):
        return basename == rule.pattern
    rel = command_path.relative_to(command_dir).as_posix()
    return fnmatch.fnmatch(basename, rule.pattern) or fnmatch.fnmatch(rel, rule.pattern)


def resolve_invocation(
    command_path: Path,
    command_dirs: list[Path],
    exec_config: ExecConfig,
) -> list[str]:
    if exec_config.malformed:
        raise ExecutionError(
            exec_config.malformed_reason or "malformed exec rules",
            status=500,
        )

    rules = exec_config.rules
    if os.access(command_path, os.X_OK) and command_path.is_file():
        return [str(command_path)]

    command_dir: Path | None = None
    for d in command_dirs:
        try:
            command_path.relative_to(d)
            command_dir = d
            break
        except ValueError:
            continue

    if command_dir is None:
        command_dir = command_path.parent

    for rule in rules:
        if _rule_matches(rule, command_path, command_dir):
            if rule.interpreter == "__wash_missing_interpreter__":
                raise ExecutionError(
                    f"unresolved interpreter for command {command_path.name}",
                    status=500,
                )
            return [rule.interpreter, *rule.extra_args, str(command_path)]

    if command_path.is_file():
        return [str(command_path)]

    raise ExecutionError(
        f"could not resolve interpreter for command {command_path.name}",
        status=500,
    )


def _stderr_merge_enabled(stage: CommandStage) -> bool:
    return stage.stderr_merge_boundary or stage.metadata.stderr_mode == "merge"


def _run_process(
    argv: list[str],
    *,
    cwd: Path,
    stdin_data: bytes | None,
    merge_stderr: bool,
) -> tuple[int, bytes, bytes]:
    stderr_opt = subprocess.STDOUT if merge_stderr else subprocess.PIPE
    proc = subprocess.Popen(
        argv,
        cwd=str(cwd),
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=stderr_opt,
        env={**os.environ, "PATH": os.environ.get("PATH", "")},
    )
    assert proc.stdin is not None
    if stdin_data is not None:
        proc.stdin.write(stdin_data)
    proc.stdin.close()

    stdout_chunks: list[bytes] = []
    stderr_chunks: list[bytes] = []

    if merge_stderr:
        assert proc.stdout is not None
        stdout_chunks.append(proc.stdout.read())
    else:
        sel = selectors.DefaultSelector()
        if proc.stdout is not None:
            sel.register(proc.stdout, selectors.EVENT_READ, "stdout")
        if proc.stderr is not None:
            sel.register(proc.stderr, selectors.EVENT_READ, "stderr")
        while sel.get_map():
            for key, _ in sel.select(timeout=0.1):
                chunk = key.fileobj.read(65536)  # type: ignore[union-attr]
                if not chunk:
                    sel.unregister(key.fileobj)
                    continue
                if key.data == "stdout":
                    stdout_chunks.append(chunk)
                else:
                    stderr_chunks.append(chunk)
        if proc.stdout is not None:
            rest = proc.stdout.read()
            if rest:
                stdout_chunks.append(rest)
        if proc.stderr is not None:
            rest = proc.stderr.read()
            if rest:
                stderr_chunks.append(rest)

    exit_code = proc.wait()
    return exit_code, b"".join(stdout_chunks), b"".join(stderr_chunks)


def execute_raw_command(
    raw: RawCommandParse,
    *,
    root: Path,
    command_dirs: list[Path],
    exec_config: ExecConfig,
    method: str,
    body: bytes,
) -> PipelineResult:
    stage = raw.stage
    invocation = resolve_invocation(stage.command_path, command_dirs, exec_config)
    invocation = invocation + [raw.raw_suffix]
    stdin_data = body if body else None
    exit_code, stdout, stderr = _run_process(
        invocation,
        cwd=root,
        stdin_data=stdin_data,
        merge_stderr=_stderr_merge_enabled(stage),
    )
    http_status = map_exit_status(stage.metadata, exit_code)
    stage_result = StageResult(
        name=stage.name,
        exit_code=exit_code,
        stdout=stdout,
        stderr=stderr,
        http_status=http_status,
    )
    content_type = stage.metadata.mime or "text/plain"
    return PipelineResult(
        stdout=stdout,
        stderr=stderr,
        content_type=content_type,
        stages=[stage_result],
        pipeline_description=f"{stage.name} {raw.raw_suffix}".strip(),
        source_path=None,
        final_command=stage.name,
        failing_stage=stage_result if http_status >= 400 else None,
        http_status=http_status,
    )


def execute_pipeline(
    pipeline: PipelineParse,
    *,
    root: Path,
    command_dirs: list[Path],
    exec_config: ExecConfig,
    body: bytes,
    symlink_policy: str = "reject-escaping",
    case_sensitive: bool = True,
) -> PipelineResult:
    stages = pipeline.stages
    if not stages:
        raise ExecutionError("empty pipeline")

    stdin_data: bytes | None = None
    if pipeline.input_suffix_raw:
        try:
            stdin_data = implied_cat_bytes(
                root,
                pipeline.input_suffix_raw,
                symlink_policy=symlink_policy,
                case_sensitive=case_sensitive,
            )
        except FileNotFoundError as exc:
            raise ExecutionError(str(exc), status=404) from exc
        except IsADirectoryError as exc:
            raise ExecutionError(str(exc), status=400) from exc
    elif body:
        stdin_data = body

    data_flow_stages = list(reversed(stages))
    stage_results: list[StageResult] = []
    current_input = stdin_data

    for i, stage in enumerate(data_flow_stages):
        invocation = resolve_invocation(stage.command_path, command_dirs, exec_config)
        invocation = invocation + stage.argv
        merge = _stderr_merge_enabled(stage)
        exit_code, stdout, stderr = _run_process(
            invocation,
            cwd=root,
            stdin_data=current_input,
            merge_stderr=merge,
        )
        http_status = map_exit_status(stage.metadata, exit_code)
        result = StageResult(
            name=stage.name,
            exit_code=exit_code,
            stdout=stdout,
            stderr=stderr,
            http_status=http_status,
        )
        stage_results.append(result)
        current_input = stdout

    first_fail = _first_failing_in_url_order(stages, stage_results)
    if first_fail is not None:
        return PipelineResult(
            stdout=first_fail.stdout,
            stderr=first_fail.stderr,
            content_type=_final_content_type(stages),
            stages=stage_results,
            pipeline_description=pipeline.pipeline_description,
            source_path=pipeline.source_path,
            final_command=stages[0].name if stages else None,
            failing_stage=first_fail,
            http_status=first_fail.http_status,
        )

    final_stage = stages[0]
    content_type = final_stage.metadata.mime or "text/plain"
    return PipelineResult(
        stdout=current_input or b"",
        stderr=b"",
        content_type=content_type,
        stages=stage_results,
        pipeline_description=pipeline.pipeline_description,
        source_path=pipeline.source_path,
        final_command=stages[0].name if stages else None,
        failing_stage=None,
        http_status=200,
    )


def _first_failing_in_url_order(
    url_stages: list[CommandStage],
    data_flow_results: list[StageResult],
) -> StageResult | None:
    by_name = {r.name: r for r in data_flow_results}
    for stage in url_stages:
        result = by_name.get(stage.name)
        if result and result.http_status >= 400:
            return result
    return None


def _final_content_type(stages: list[CommandStage]) -> str:
    if stages:
        return stages[0].metadata.mime or "text/plain"
    return "text/plain"

File: impls/dart/lib/executor/executor.dart
import 'dart:io';
import 'dart:async';

import '../filesystem/filesystem.dart';
import '../metadata/metadata.dart';
import '../parser/parser.dart';
import 'package:glob/glob.dart';

class ExecutionError implements Exception {
  final String message;
  final int status;

  ExecutionError(this.message, {this.status = 500});

  @override
  String toString() => 'ExecutionError($status): $message';
}

class ExecRule {
  final String pattern;
  final String interpreter;
  final List<String> extraArgs;

  ExecRule(
      {required this.pattern,
      required this.interpreter,
      required this.extraArgs});
}

class StageResult {
  final String name;
  final int exitCode;
  final List<int> stdout;
  final List<int> stderr;
  final int httpStatus;

  StageResult({
    required this.name,
    required this.exitCode,
    required this.stdout,
    required this.stderr,
    required this.httpStatus,
  });
}

class PipelineResult {
  final List<int> stdout;
  final List<int> stderr;
  final String contentType;
  final List<StageResult> stages;
  final String pipelineDescription;
  final String? sourcePath;
  final String? finalCommand;
  final StageResult? failingStage;
  final int httpStatus;

  PipelineResult({
    required this.stdout,
    required this.stderr,
    required this.contentType,
    required this.stages,
    required this.pipelineDescription,
    required this.sourcePath,
    this.finalCommand,
    this.failingStage,
    this.httpStatus = 200,
  });
}

class ExecConfig {
  final List<ExecRule> rules;
  final bool malformed;
  final String? malformedReason;

  ExecConfig(this.rules, {this.malformed = false, this.malformedReason});
}

ExecConfig loadExecRules(String root) {
  final execFile = File('$root${Platform.pathSeparator}exec');
  final rules = <ExecRule>[];
  if (!execFile.existsSync()) return ExecConfig(rules);

  for (final line in execFile.readAsLinesSync()) {
    final stripped = line.trim();
    if (stripped.isEmpty || stripped.startsWith('#')) continue;
    final parts = stripped.split(RegExp(r'\s+'));
    if (parts.length < 2) {
      return ExecConfig(
        rules,
        malformed: true,
        malformedReason: 'malformed exec rule: $line',
      );
    }
    rules.add(ExecRule(
      pattern: parts[0],
      interpreter: parts[1],
      extraArgs: parts.sublist(2),
    ));
  }
  return ExecConfig(rules);
}

bool _hasGlob(String pattern) {
  return pattern.contains('*') ||
      pattern.contains('?') ||
      pattern.contains('[');
}

bool _ruleMatches(ExecRule rule, String commandPath, String commandDir) {
  final basename = commandPath.split(Platform.pathSeparator).last;
  if (!_hasGlob(rule.pattern)) {
    return basename == rule.pattern;
  }
  String rel;
  try {
    rel = commandPath
        .replaceFirst(commandDir, '')
        .replaceFirst(RegExp(r'^[/\\]'), '');
    rel = rel.replaceAll('\\', '/');
  } catch (_) {
    rel = basename;
  }
  final globBasename = Glob(rule.pattern);
  final globRel = Glob(rule.pattern);
  return globBasename.matches(basename) || globRel.matches(rel);
}

List<String> resolveInvocation(
  String commandPath,
  List<String> commandDirs,
  ExecConfig execConfig,
) {
  if (execConfig.malformed) {
    throw ExecutionError(
      execConfig.malformedReason ?? 'malformed exec rules',
      status: 500,
    );
  }

  final file = File(commandPath);
  if (file.existsSync()) {
    final stat = file.statSync();
    // Check executable bit (mode & 0o111)
    if ((stat.mode & 0x49) != 0) {
      return [commandPath];
    }
  }

  String? commandDir;
  for (final dir in commandDirs) {
    if (commandPath.startsWith(dir)) {
      commandDir = dir;
      break;
    }
  }
  commandDir ??=
      commandPath.substring(0, commandPath.lastIndexOf(Platform.pathSeparator));

  for (final rule in execConfig.rules) {
    if (_ruleMatches(rule, commandPath, commandDir)) {
      if (rule.interpreter == '__wash_missing_interpreter__') {
        throw ExecutionError(
          'unresolved interpreter for command ${commandPath.split(Platform.pathSeparator).last}',
          status: 500,
        );
      }
      return [rule.interpreter, ...rule.extraArgs, commandPath];
    }
  }

  if (file.existsSync()) {
    return [commandPath];
  }

  throw ExecutionError(
    'could not resolve interpreter for command ${commandPath.split(Platform.pathSeparator).last}',
    status: 500,
  );
}

bool _stderrMergeEnabled(CommandStage stage) {
  return stage.stderrMergeBoundary || stage.metadata.stderrMode == 'merge';
}

Future<(int, List<int>, List<int>)> _runProcess(
  List<String> argv, {
  required String cwd,
  required List<int>? stdinData,
  required bool mergeStderr,
}) async {
  final process = await Process.start(
    argv[0],
    argv.sublist(1),
    workingDirectory: cwd,
    environment: {
      ...Platform.environment,
    },
  );

  if (stdinData != null && stdinData.isNotEmpty) {
    process.stdin.add(stdinData);
  }
  await process.stdin.close();

  final stdoutFuture =
      process.stdout.fold<List<int>>([], (prev, chunk) => [...prev, ...chunk]);
  if (mergeStderr) {
    // stderr is already merged into stdout by redirecting stderr to stdout
    // We need a different approach: use process.stderr and interleave
    // For merge, we collect stderr separately but append to stdout result
    final stderrFuture = process.stderr
        .fold<List<int>>([], (prev, chunk) => [...prev, ...chunk]);
    final exitCode = await process.exitCode;
    final stdoutBytes = await stdoutFuture;
    final stderrBytes = await stderrFuture;
    return (exitCode, [...stdoutBytes, ...stderrBytes], <int>[]);
  } else {
    final stderrFuture = process.stderr
        .fold<List<int>>([], (prev, chunk) => [...prev, ...chunk]);
    final exitCode = await process.exitCode;
    final stdoutBytes = await stdoutFuture;
    final stderrBytes = await stderrFuture;
    return (exitCode, stdoutBytes, stderrBytes);
  }
}

Future<PipelineResult> executeRawCommand(
  RawCommandParse raw, {
  required String root,
  required List<String> commandDirs,
  required ExecConfig execConfig,
  required String method,
  required List<int> body,
}) async {
  final stage = raw.stage;
  final invocation =
      resolveInvocation(stage.commandPath, commandDirs, execConfig);
  final fullInvocation = [...invocation, raw.rawSuffix];
  final stdinData = body.isNotEmpty ? body : null;
  final (exitCode, stdout, stderr) = await _runProcess(
    fullInvocation,
    cwd: root,
    stdinData: stdinData,
    mergeStderr: _stderrMergeEnabled(stage),
  );
  final httpStatus = mapExitStatus(stage.metadata, exitCode);
  final stageResult = StageResult(
    name: stage.name,
    exitCode: exitCode,
    stdout: stdout,
    stderr: stderr,
    httpStatus: httpStatus,
  );
  final contentType = stage.metadata.mime ?? 'text/plain';
  return PipelineResult(
    stdout: stdout,
    stderr: stderr,
    contentType: contentType,
    stages: [stageResult],
    pipelineDescription: '${stage.name} ${raw.rawSuffix}'.trim(),
    sourcePath: null,
    finalCommand: stage.name,
    failingStage: httpStatus >= 400 ? stageResult : null,
    httpStatus: httpStatus,
  );
}

Future<PipelineResult> executePipeline(
  PipelineParse pipeline, {
  required String root,
  required List<String> commandDirs,
  required ExecConfig execConfig,
  required List<int> body,
  String symlinkPolicy = 'reject-escaping',
  bool caseSensitive = true,
}) async {
  final stages = pipeline.stages;
  if (stages.isEmpty) {
    throw ExecutionError('empty pipeline');
  }

  List<int>? stdinData;
  if (pipeline.inputSuffixRaw != null && pipeline.inputSuffixRaw!.isNotEmpty) {
    try {
      stdinData = impliedCatBytes(
        root,
        pipeline.inputSuffixRaw!,
        symlinkPolicy: symlinkPolicy,
        caseSensitive: caseSensitive,
      );
    } on FileSystemException catch (e) {
      final msg = e.message.toLowerCase();
      if (msg.contains('not found') || msg.contains('no such')) {
        throw ExecutionError(e.message, status: 404);
      }
      if (msg.contains('directory')) {
        throw ExecutionError(e.message, status: 400);
      }
      if (msg.contains('escapes root') || msg.contains('not permitted')) {
        throw ExecutionError(e.message, status: 403);
      }
      throw ExecutionError(e.message, status: 404);
    } on NameEscapeError {
      throw ExecutionError('path not permitted', status: 403);
    } on NameLoopError {
      throw ExecutionError('name resolution loop detected', status: 508);
    } on RootEscapeError {
      throw ExecutionError('path not permitted', status: 403);
    } on SymlinkEscapeError {
      throw ExecutionError('path not permitted', status: 403);
    }
  } else if (body.isNotEmpty) {
    stdinData = body;
  }

  final dataFlowStages = stages.reversed.toList();
  final stageResults = <StageResult>[];
  List<int>? currentInput = stdinData;

  for (final stage in dataFlowStages) {
    final invocation =
        resolveInvocation(stage.commandPath, commandDirs, execConfig);
    final fullInvocation = [...invocation, ...stage.argv];
    final merge = _stderrMergeEnabled(stage);
    final (exitCode, stdout, stderr) = await _runProcess(
      fullInvocation,
      cwd: root,
      stdinData: currentInput,
      mergeStderr: merge,
    );
    final httpStatus = mapExitStatus(stage.metadata, exitCode);
    final result = StageResult(
      name: stage.name,
      exitCode: exitCode,
      stdout: stdout,
      stderr: stderr,
      httpStatus: httpStatus,
    );
    stageResults.add(result);
    currentInput = stdout;
  }

  final firstFail = _firstFailingInUrlOrder(stages, stageResults);
  if (firstFail != null) {
    return PipelineResult(
      stdout: firstFail.stdout,
      stderr: firstFail.stderr,
      contentType: _finalContentType(stages),
      stages: stageResults,
      pipelineDescription: pipeline.pipelineDescription,
      sourcePath: pipeline.sourcePath,
      finalCommand: stages.isNotEmpty ? stages[0].name : null,
      failingStage: firstFail,
      httpStatus: firstFail.httpStatus,
    );
  }

  final finalStage = stages[0];
  final contentType = finalStage.metadata.mime ?? 'text/plain';
  return PipelineResult(
    stdout: currentInput ?? [],
    stderr: [],
    contentType: contentType,
    stages: stageResults,
    pipelineDescription: pipeline.pipelineDescription,
    sourcePath: pipeline.sourcePath,
    finalCommand: stages.isNotEmpty ? stages[0].name : null,
    failingStage: null,
    httpStatus: 200,
  );
}

StageResult? _firstFailingInUrlOrder(
  List<CommandStage> urlStages,
  List<StageResult> dataFlowResults,
) {
  final byName = <String, StageResult>{};
  for (final r in dataFlowResults) {
    byName[r.name] = r;
  }
  for (final stage in urlStages) {
    final result = byName[stage.name];
    if (result != null && result.httpStatus >= 400) {
      return result;
    }
  }
  return null;
}

String _finalContentType(List<CommandStage> stages) {
  if (stages.isNotEmpty) {
    return stages[0].metadata.mime ?? 'text/plain';
  }
  return 'text/plain';
}
